CondTools/SiPhase2Tracker/plugins/DTCCablingMapProducer.cc

```diff
@@ -0,0 +1,299 @@
+	csvFormat_idtcid_                        (iConfig.getParameter<unsigned>("csvFormat_idtcid"  )),
+	iovBeginTime_                            (iConfig.getParameter<long long unsigned int>("iovBeginTime")),
+	pCablingMap_                             (std::make_unique<TrackerDetToDTCELinkCablingMap>()),
+	record_                                  (iConfig.getParameter<std::string>("record"))
+{
+	LoadModulesToDTCCablingMapFromCSV(iConfig.getParameter<std::vector<std::string>>("modulesToDTCCablingCSVFileNames"));
+}
+
+void DTCCablingMapProducer::beginJob()
+{
+	
+}
+
+void DTCCablingMapProducer::LoadModulesToDTCCablingMapFromCSV(std::vector<std::string> const& modulesToDTCCablingCSVFileNames)
+{
+	using namespace std;
+	
+	for (std::string const& csvFileName : modulesToDTCCablingCSVFileNames)
+	{
+		edm::FileInPath csvFilePath(csvFileName);
+		
+		ifstream csvFile;
+		csvFile.open(csvFilePath.fullPath().c_str());
+		
+		if (csvFile.is_open())
+		{
+			string csvLine;
+			
+			unsigned lineNumber = 0;
+			
+			while (std::getline(csvFile, csvLine))
+			{
+				if (verbosity_ >= 1)
+				{
+					edm::LogInfo("CSVParser") << "Reading CSV file line: " << ++lineNumber << ": \"" << csvLine << "\"" << endl;
+				}
+				
+				istringstream csvStream(csvLine);
+				vector<string> csvColumn;
+				string csvElement;
+				
+				while (std::getline(csvStream, csvElement, ','))
+				{
+					trim(csvElement);
+					csvColumn.push_back(csvElement);
+				}
+				
+				if (verbosity_ >= 2)
+				{
+					ostringstream splitted_line_info;
+					
+					splitted_line_info << "-- split line is: [";
+					
+					for (string const& s : csvColumn)
+						splitted_line_info << "\"" << s << "\", ";
+					
+					splitted_line_info << "]" << endl;
+					
+					edm::LogInfo("CSVParser") << splitted_line_info.str();
+				}
+				
+				if (csvColumn.size() == csvFormat_ncolumns_)
+				{
+					// Skip the legend lines
+					if (0 == csvColumn[0].compare(std::string("Module DetId/U")))
+					{
+						if (verbosity_ >= 1)
+						{
+							edm::LogInfo("CSVParser") << "-- skipping legend line" << endl;
+						}
+						continue;
+					}
+					
+					uint32_t detIdRaw;
+					
+					try
+					{
+						detIdRaw = std::stoi( csvColumn[csvFormat_idetid_] );
+					}
+					catch (std::exception e)
+					{
+						if (verbosity_ >= 0)
+						{
+							edm::LogError("CSVParser") << "-- malformed DetId string in CSV file: \"" << csvLine << "\"" << endl;
+						}
+						throw e;
+					}
+					
+					unsigned const dtc_id = strtoul(csvColumn[csvFormat_idtcid_].c_str(), nullptr, 10);
+					unsigned gbt_id ;
+					unsigned elink_id;
+					
+					if (generate_fake_valid_gbtlink_and_elinkid_)
+					{
+						for (gbt_id = 0; gbt_id < 72; ++gbt_id)
+						{
+							for  (elink_id = 0; elink_id < 7; ++elink_id)
+							{
+								if ( !(pCablingMap_->knowsDTCELinkId( DTCELinkId(dtc_id, gbt_id, elink_id) )) )
+									goto gbtlink_and_elinkid_generator_end;
```

Comment:
Please do not use the `goto` statement.

Reply:
One of the last meaningful uses of the goto statement is to break out of nested loops. I used it there since it is explicit what it is meant to do, and translates directly to a jump instruction. 

Can you suggest an elegant alternative?

Reply:
Hmmm, with careful check the CMSSW coding rule, this is one of the exception to use `goto`.

Thank you, I learned something.